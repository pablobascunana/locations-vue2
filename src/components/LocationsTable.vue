<template>
  <div>
    <v-data-table
      :headers="headers"
      :items="markers"
      sort-by="title"
    >
      <template v-slot:top>
        <v-toolbar flat>
          <v-toolbar-title>{{ $t('locations.tableTitle') }}</v-toolbar-title>
          <v-spacer></v-spacer>
          <v-btn
            id="showAddLocation"
            color="primary"
            dark
            @click="addLocation"
          >{{ $t('generic.buttons.add') }}
          </v-btn>
        </v-toolbar>
      </template>
      <template v-slot:[`item.actions`]="{ item }">
        <v-tooltip left>
          <template v-slot:activator="{ on }">
            <v-icon
              id="deleteLocation"
              small
              v-on="on"
              @click="deleteLocation(item)"
            >mdi-delete
            </v-icon>
          </template>
          <span>{{ $t('locations.delete') }}</span>
        </v-tooltip>
      </template>
    </v-data-table>
    <RemoveLocationDialog
      :marker="markerToDelete"
      :userUuid="userUuid"
      @updateMarkers="updateMarkersHandler"
    ></RemoveLocationDialog>
  </div>
</template>

<script>
export default {
  name: 'locationsTable',
  props: {
    markers: { type: Array, default: () => []},
    userUuid: { type: String, required: true }
  },
  components: {
    RemoveLocationDialog: () => import('@/components/dialogs/RemoveLocation')
  },
  data() {
    return {
      headers: [
        { text: 'Localizaci√≥n', value: 'title', align: 'center' },
        { text: 'Latitud', value: 'position.lat', align: 'center', sortable: false},
        { text: 'Longitud', value: 'position.lng', align: 'center', sortable: false},
        { text: 'Fecha', value: 'date', align: 'center', sortable: true },
        { text: 'Acciones', value: 'actions', align: 'center', sortable: false }
      ],
      markerToDelete: {}
    }
  },
  methods: {
    addLocation() {
      this.$store.commit('application/isAddLocationDialogVisible', true);
    },
    async deleteLocation(marker) {
      this.markerToDelete = marker;
      this.$store.commit('application/isRemoveLocationDialogVisible', true); 
    },
    updateMarkersHandler(marker) {
      this.$emit('updateMarkers', marker);
    }
  }
}
</script>
