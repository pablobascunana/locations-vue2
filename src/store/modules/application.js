function initialApplicationState() {
  return {
    isLogin: true,
    accessToken: undefined,
    refreshToken: undefined,
    addLocationDialogVisible: false,
    removeLocationDialogVisible: false
  }
}

const application = {
  namespaced: true,
  state: initialApplicationState,
  mutations: {
    resetApplicationStore(state) {
      const initState = initialApplicationState();
      Object.keys(initState).forEach(key => {
        state[key] = initState[key];
      })
    },
    isLogin(state, visible) {
      state.isLogin = visible;
    },
    setAccessToken(state, token) {
      state.accessToken = `Bearer ${token}`;
    },
    setRefreshToken(state, token) {
      state.refreshToken = token;
    },
    isAddLocationDialogVisible(state, visible) {
      state.addLocationDialogVisible = visible;
    },
    isRemoveLocationDialogVisible(state, visible) {
      state.removeLocationDialogVisible = visible
    }
  }
};

export default application;
  