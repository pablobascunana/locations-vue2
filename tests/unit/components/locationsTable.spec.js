import { createLocalVue, shallowMount } from '@vue/test-utils'

import Vuetify from 'vuetify'
import Vuex from 'vuex'

import application from '@/store/modules/application.js'

import LocationsTable from '@/components/LocationsTable.vue'

import * as historicalLocationsMock from '../mocks/historicalLocations.mock.js'

const localVue = createLocalVue();
localVue.use(Vuex);
let vuetify = new Vuetify();

describe('Locations Table', () => {
    let wrapper;
    let store;
    let state = {};
    let actions = {};
    let modules = {
        application
    };
    beforeEach(() => {
        store = new Vuex.Store({ state, actions, modules });

        wrapper = shallowMount(LocationsTable, {
            localVue,
            vuetify,
            store,
            mocks:Â {
                $t: () => {'some specific text in any language'}
            }
        });
    });

    it("Should match snapshot", () => {
        expect(wrapper).toMatchSnapshot()    
    });

    it('Renders a vue instance', () => {
        expect(wrapper.isVueInstance()).toBe(true);
    });

    it('Call to add location function', () => {
        const addLocationFn = jest.spyOn(wrapper.vm, 'addLocation');
        wrapper.vm.addLocation();
        expect(addLocationFn).toBeCalled();
        expect(wrapper.vm.$store.state.application.addLocationDialogVisible).toBe(true);
    });

    it('Call to delete location function', () => {
        const deleteLocationFn = jest.spyOn(wrapper.vm, 'deleteLocation');
        wrapper.vm.deleteLocation(historicalLocationsMock.MARKER);
        expect(deleteLocationFn).toBeCalled();
        expect(wrapper.vm.markerToDelete).toStrictEqual(historicalLocationsMock.MARKER);
        expect(wrapper.vm.$store.state.application.removeLocationDialogVisible).toBe(true);
    });

    it('Call to update markers handler function', () => {
        const updateMarkersHandlerFn = jest.spyOn(wrapper.vm, 'updateMarkersHandler');
        wrapper.vm.updateMarkersHandler(historicalLocationsMock.MARKER);
        expect(updateMarkersHandlerFn).toBeCalled();
    });
});
