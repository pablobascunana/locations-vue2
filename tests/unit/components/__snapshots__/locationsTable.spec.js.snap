// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locations Table Should match snapshot 1`] = `
<div>
  <v-data-table headers="[object Object],[object Object],[object Object],[object Object],[object Object]" items="" sort-by="title"></v-data-table>
  <v-dialog max-width="500px">
    <v-card>
      <v-card-title>

        <v-spacer></v-spacer>
        <v-icon id="close" aria-label="Close">mdi-close</v-icon>
      </v-card-title>
      <v-card-text class="pt-0 pb-0 text-center">
        <p><span class="font-weight-bold"> </span>?</p>
      </v-card-text>
      <v-card-actions>
        <v-spacer></v-spacer>
        <v-btn id="removeLocation" color="red" text="">
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</div>
`;
